File: android/app/voice/VoiceCommandManager.kt
package com.gemnav.app.voice

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Central coordinator for voice command functionality.
 * Manages speech recognition, command parsing, execution, and TTS responses.
 */
@Singleton
class VoiceCommandManager @Inject constructor(
    private val speechRecognitionService: SpeechRecognitionService,
    private val commandParser: CommandParser,
    private val commandExecutor: CommandExecutor,
    private val voiceResponseService: VoiceResponseService,
    private val wakeWordDetector: WakeWordDetector?,
    private val tier: SubscriptionTier,
    private val coroutineScope: CoroutineScope
) {
    private val _voiceState = MutableStateFlow<VoiceState>(VoiceState.Idle)
    val voiceState: StateFlow<VoiceState> = _voiceState.asStateFlow()
    
    private val _lastTranscript = MutableStateFlow<String?>(null)
    val lastTranscript: StateFlow<String?> = _lastTranscript.asStateFlow()
    
    private val conversationContext = mutableListOf<ConversationTurn>()
    private var isWakeWordEnabled = false
    
    init {
        setupSpeechRecognition()
        if (tier.allowsWakeWord()) {
            setupWakeWord()
        }
    }
    
    private fun setupSpeechRecognition() {
        speechRecognitionService.setOnTranscriptListener { transcript ->
            coroutineScope.launch {
                handleTranscript(transcript)
            }
        }
        
        speechRecognitionService.setOnErrorListener { error ->
            handleRecognitionError(error)
        }
    }
    
    private fun setupWakeWord() {
        wakeWordDetector?.setOnWakeWordDetectedListener {
            if (isWakeWordEnabled && _voiceState.value == VoiceState.Idle) {
                startListening(VoiceTrigger.WakeWord)
            }
        }
    }
    
    /**
     * Start listening for voice input
     */
    fun startListening(trigger: VoiceTrigger = VoiceTrigger.Manual) {
        if (_voiceState.value == VoiceState.Listening) return
        
        _voiceState.value = VoiceState.Listening
        speechRecognitionService.startListening()
    }
    
    /**
     * Stop listening for voice input
     */
    fun stopListening() {
        if (_voiceState.value != VoiceState.Listening) return
        
        speechRecognitionService.stopListening()
        _voiceState.value = VoiceState.Idle
    }
    
    /**
     * Process voice transcript
     */
    private suspend fun handleTranscript(transcript: String) {
        _lastTranscript.value = transcript
        _voiceState.value = VoiceState.Processing
        
        try {
            // Add to conversation context if Plus/Pro
            if (tier.allowsAdvancedVoice()) {
                conversationContext.add(
                    ConversationTurn(role = "user", content = transcript)
                )
            }
            
            // Parse command using Gemini
            val command = commandParser.parse(transcript, conversationContext)
            
            // Execute command
            val result = commandExecutor.execute(command)
            
            // Handle result
            when (result) {
                is CommandResult.Success -> {
                    speak(result.message, interrupt = true)
                    if (tier.allowsAdvancedVoice()) {
                        conversationContext.add(
                            ConversationTurn(role = "assistant", content = result.message)
                        )
                    }
                }
                
                is CommandResult.Error -> {
                    speak(result.message, interrupt = true)
                    _voiceState.value = VoiceState.Error(result.message)
                }
                
                is CommandResult.TierRestricted -> {
                    speak(result.message, interrupt = true)
                    _voiceState.value = VoiceState.Error(result.message)
                }
                
                is CommandResult.NeedsClarification -> {
                    speak(result.question, interrupt = true)
                    // Keep listening for follow-up
                    startListening(VoiceTrigger.Continuation)
                }
            }
            
        } catch (e: Exception) {
            val errorMessage = "Sorry, I couldn't process that command"
            _voiceState.value = VoiceState.Error(errorMessage)
            speak(errorMessage, interrupt = true)
        }
    }
    
    /**
     * Speak text via TTS
     */
    fun speak(text: String, interrupt: Boolean = false) {
        _voiceState.value = VoiceState.Speaking(text)
        voiceResponseService.speak(text, interrupt)
        
        // Auto-return to idle after speaking
        coroutineScope.launch {
            kotlinx.coroutines.delay(text.length * 50L) // Rough estimate
            if (_voiceState.value is VoiceState.Speaking) {
                _voiceState.value = VoiceState.Idle
            }
        }
    }
    
    /**
     * Cancel current speech
     */
    fun cancelSpeech() {
        voiceResponseService.stop()
        if (_voiceState.value is VoiceState.Speaking) {
            _voiceState.value = VoiceState.Idle
        }
    }
    
    /**
     * Enable/disable wake word detection (Plus/Pro only)
     */
    fun setWakeWordEnabled(enabled: Boolean) {
        if (!tier.allowsWakeWord()) {
            return
        }
        
        isWakeWordEnabled = enabled
        if (enabled) {
            wakeWordDetector?.start()
        } else {
            wakeWordDetector?.stop()
        }
    }
    
    /**
     * Clear conversation context
     */
    fun clearConversationContext() {
        conversationContext.clear()
    }
    
    /**
     * Handle recognition errors
     */
    private fun handleRecognitionError(error: SpeechRecognitionError) {
        val message = when (error) {
            SpeechRecognitionError.NO_MATCH -> "I didn't catch that. Please try again."
            SpeechRecognitionError.NETWORK_ERROR -> "Network error. Voice recognition unavailable."
            SpeechRecognitionError.PERMISSION_DENIED -> "Microphone permission required for voice commands."
            SpeechRecognitionError.INSUFFICIENT_PERMISSIONS -> "Microphone permission required for voice commands."
            SpeechRecognitionError.SERVICE_UNAVAILABLE -> "Speech recognition service unavailable."
            else -> "Voice recognition error. Please try again."
        }
        
        _voiceState.value = VoiceState.Error(message)
        speak(message, interrupt = true)
    }
    
    /**
     * Cleanup resources
     */
    fun shutdown() {
        stopListening()
        wakeWordDetector?.stop()
        voiceResponseService.shutdown()
    }
}

// Voice states
sealed class VoiceState {
    object Idle : VoiceState()
    object Listening : VoiceState()
    object Processing : VoiceState()
    data class Speaking(val text: String) : VoiceState()
    data class Error(val message: String) : VoiceState()
}

// Voice trigger types
enum class VoiceTrigger {
    Manual,      // User tapped microphone button
    WakeWord,    // Wake word detected
    Continuation // Continuation of multi-turn conversation
}

// Conversation context for multi-turn dialogue
data class ConversationTurn(
    val role: String,  // "user" or "assistant"
    val content: String,
    val timestamp: Long = System.currentTimeMillis()
)

// Subscription tier extensions
fun SubscriptionTier.allowsWakeWord(): Boolean {
    return this == SubscriptionTier.PLUS || this == SubscriptionTier.PRO
}

fun SubscriptionTier.allowsAdvancedVoice(): Boolean {
    return this == SubscriptionTier.PLUS || this == SubscriptionTier.PRO
}

File: ios/app/voice/VoiceCommandManager.swift
import Foundation
import Combine

/**
 * Central coordinator for voice command functionality.
 * Manages speech recognition, command parsing, execution, and TTS responses.
 */
class VoiceCommandManager: ObservableObject {
    // Published state for SwiftUI
    @Published private(set) var voiceState: VoiceState = .idle
    @Published private(set) var lastTranscript: String?
    
    private let speechRecognitionService: SpeechRecognitionService
    private let commandParser: CommandParser
    private let commandExecutor: CommandExecutor
    private let voiceResponseService: VoiceResponseService
    private let wakeWordDetector: WakeWordDetector?
    private let tier: SubscriptionTier
    
    private var conversationContext: [ConversationTurn] = []
    private var isWakeWordEnabled = false
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    
    init(
        speechRecognitionService: SpeechRecognitionService,
        commandParser: CommandParser,
        commandExecutor: CommandExecutor,
        voiceResponseService: VoiceResponseService,
        wakeWordDetector: WakeWordDetector?,
        tier: SubscriptionTier
    ) {
        self.speechRecognitionService = speechRecognitionService
        self.commandParser = commandParser
        self.commandExecutor = commandExecutor
        self.voiceResponseService = voiceResponseService
        self.wakeWordDetector = wakeWordDetector
        self.tier = tier
        
        setupSpeechRecognition()
        
        if tier.allowsWakeWord {
            setupWakeWord()
        }
    }
    
    // MARK: - Setup
    
    private func setupSpeechRecognition() {
        speechRecognitionService.setOnTranscriptListener { [weak self] transcript in
            Task {
                await self?.handleTranscript(transcript)
            }
        }
        
        speechRecognitionService.setOnErrorListener { [weak self] error in
            self?.handleRecognitionError(error)
        }
    }
    
    private func setupWakeWord() {
        wakeWordDetector?.setOnWakeWordDetectedListener { [weak self] in
            guard let self = self else { return }
            if self.isWakeWordEnabled && self.voiceState == .idle {
                self.startListening(trigger: .wakeWord)
            }
        }
    }
    
    // MARK: - Voice Control
    
    func startListening(trigger: VoiceTrigger = .manual) {
        guard voiceState != .listening else { return }
        
        voiceState = .listening
        
        do {
            try speechRecognitionService.startListening()
        } catch {
            voiceState = .error(message: "Failed to start listening: \(error.localizedDescription)")
        }
    }
    
    func stopListening() {
        guard voiceState == .listening else { return }
        
        speechRecognitionService.stopListening()
        voiceState = .idle
    }
    
    // MARK: - Transcript Processing
    
    private func handleTranscript(_ transcript: String) async {
        await MainActor.run {
            lastTranscript = transcript
            voiceState = .processing
        }
        
        // Add to conversation context if Plus/Pro
        if tier.allowsAdvancedVoice {
            conversationContext.append(
                ConversationTurn(role: "user", content: transcript)
            )
        }
        
        // Parse command using Gemini
        let command = await commandParser.parse(transcript: transcript, context: conversationContext)
        
        // Execute command
        let result = await commandExecutor.execute(command: command)
        
        // Handle result
        await MainActor.run {
            handleCommandResult(result)
        }
    }
    
    private func handleCommandResult(_ result: CommandResult) {
        switch result {
        case .success(let message, _):
            speak(text: message, interrupt: true)
            if tier.allowsAdvancedVoice {
                conversationContext.append(
                    ConversationTurn(role: "assistant", content: message)
                )
            }
            
        case .error(let message):
            speak(text: message, interrupt: true)
            voiceState = .error(message: message)
            
        case .tierRestricted(let message):
            speak(text: message, interrupt: true)
            voiceState = .error(message: message)
            
        case .needsClarification(let question):
            speak(text: question, interrupt: true)
            // Keep listening for follow-up
            startListening(trigger: .continuation)
        }
    }
    
    // MARK: - Speech Output
    
    func speak(text: String, interrupt: Bool = false) {
        voiceState = .speaking(text: text)
        voiceResponseService.speak(text: text, interrupt: interrupt)
        
        // Auto-return to idle after speaking
        // Rough estimate: 50ms per character
        let delay = Double(text.count) * 0.05
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self else { return }
            if case .speaking = self.voiceState {
                self.voiceState = .idle
            }
        }
    }
    
    func cancelSpeech() {
        voiceResponseService.stop()
        if case .speaking = voiceState {
            voiceState = .idle
        }
    }
    
    // MARK: - Wake Word
    
    func setWakeWordEnabled(_ enabled: Bool) {
        guard tier.allowsWakeWord else { return }
        
        isWakeWordEnabled = enabled
        if enabled {
            wakeWordDetector?.start()
        } else {
            wakeWordDetector?.stop()
        }
    }
    
    // MARK: - Context Management
    
    func clearConversationContext() {
        conversationContext.removeAll()
    }
    
    // MARK: - Error Handling
    
    private func handleRecognitionError(_ error: SpeechRecognitionError) {
        let message: String
        
        switch error {
        case .noMatch:
            message = "I didn't catch that. Please try again."
        case .networkError:
            message = "Network error. Voice recognition unavailable."
        case .permissionDenied, .insufficientPermissions:
            message = "Microphone permission required for voice commands."
        case .serviceUnavailable:
            message = "Speech recognition service unavailable."
        case .recognizerBusy:
            message = "Speech recognizer is busy. Please try again."
        case .audioError:
            message = "Audio system error occurred."
        case .unknown:
            message = "Voice recognition error. Please try again."
        }
        
        voiceState = .error(message: message)
        speak(text: message, interrupt: true)
    }
    
    // MARK: - Cleanup
    
    func shutdown() {
        stopListening()
        wakeWordDetector?.stop()
        voiceResponseService.shutdown()
        cancellables.removeAll()
    }
    
    deinit {
        shutdown()
    }
}
